src/sidechaindb.cpp

```diff
@@ -106,148 +134,172 @@ std::vector<SidechainDeposit> SidechainDB::GetDeposits(uint8_t nSidechain) const
+        return true;
+    else
+    if (SCDB[SIDECHAIN_WIMBLE].IsPopulated())
+        return true;
 
-    SCDBIndex& index = SCDB[nSidechain];
+    return false;
+}
 
-    // Just checking if member can be inserted
-    if (fJustCheck && !index.IsFull())
-        return true;
+bool SidechainDB::HaveDepositCached(const SidechainDeposit &deposit) const
+{
+    for (const SidechainDeposit& d : vDepositCache) {
+        if (d == deposit)
+            return true;
+    }
+    return false;
+}
+
+bool SidechainDB::HaveWTJoinCached(const uint256& wtxid) const
+{
+    for (const CTransaction& tx : vWTJoinCache) {
+        if (tx.GetHash() == wtxid)
+            return true;
+    }
+    return false;
+}
+
+void SidechainDB::Reset()
+{
+    // Clear out SCDB
+    for (const Sidechain& s : ValidSidechains)
+        SCDB[s.nSidechain].ClearMembers();
+
+    // Clear out LD
+    mapBMMLD.clear();
+    std::queue<uint256> queueEmpty;
+    std::swap(queueBMMLD, queueEmpty);
+
+    // Clear out Deposit data
+    vDepositCache.clear();
 
-    // Insert member
-    return (index.InsertMember(member));
+    // Clear out cached WT^(s)
+    vWTJoinCache.clear();
+
+    // Reset hashBlockLastSeen
+    hashBlockLastSeen.SetNull();
+}
+
+std::string SidechainDB::ToString() const
+{
+    std::string str;
+    str += "SidechainDB:\n";
+    for (const Sidechain& s : ValidSidechains) {
+        // Print sidechain name
+        str += "Sidechain: " + s.GetSidechainName() + "\n";
+        // Print sidechain WT^ workscore(s)
+        std::vector<SidechainWTJoinState> vState = GetState(s.nSidechain);
+        for (const SidechainWTJoinState& state : vState) {
+            str += "WT^: " + state.wtxid.ToString() + "\n";
+            str += "workscore: " + std::to_string(state.nWorkScore) + "\n";
+        }
+        str += "\n";
+    }
+    return str;
 }
 
 bool SidechainDB::Update(int nHeight, const uint256& hashBlock, const std::vector<CTxOut>& vout, std::string& strError)
 {
     if (hashBlock.IsNull())
         return false;
+    if (!vout.size())
+        return false;
+
+    // TODO skip if nHeight < drivechains activation block height
 
     // If a sidechain's tau period ended, reset WT^ verification status
     for (const Sidechain& s : ValidSidechains) {
         if (nHeight > 0 && (nHeight % s.GetTau()) == 0)
             SCDB[s.nSidechain].ClearMembers();
     }
+    // TODO clear out cached WT^(s) that belong to the Sidechain
+    // that was just reset.
 
     /*
-     * Only one state script of the current version is valid.
-     * State scripts with invalid version numbers will be ignored.
-     * If there are multiple state scripts with valid version numbers
-     * the entire coinbase will be ignored by SCDB and a default
-     * ignore vote will be cast. If there isn't a state update in
-     * the transaction outputs, a default ignore vote will be cast.
+     * Now we will look for data that is relevent to SCDB
```

Comment:
Does it have to be every block's coinbase? Or only blocks that occur at the end of a tau period?